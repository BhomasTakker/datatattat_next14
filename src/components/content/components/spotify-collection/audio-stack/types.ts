import { SearchType } from "@/types/api/spotify";
import { CollectionItem } from "@/types/data-structures/collection/item/item";

export type SpotifyCollectionItem = CollectionItem & {
	id: string;
	media: {
		type: SearchType;
	};
};

export type AudioStackClientProps = {
	items: SpotifyCollectionItem[];
};

export type SpotifyArticleProps = {
	item: SpotifyCollectionItem;
	onClick: (index: number) => void;
	index: number;
};
