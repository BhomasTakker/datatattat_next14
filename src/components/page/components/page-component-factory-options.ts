export enum PageComponentsOptions {
	STACK = "Stack",
	GRID = "Grid",
}
