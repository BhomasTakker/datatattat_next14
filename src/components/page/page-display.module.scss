@use "../../scss/abstracts/units/" as *;

.page {
	display: flex;
	flex-direction: column;
}
