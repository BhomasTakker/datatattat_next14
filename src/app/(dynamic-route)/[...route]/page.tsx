import { getPage } from "@/actions/page/page-actions";
import { IPage } from "@/types/page";
import { PageDisplay } from "@/components/page/page-display";
import styles from "../../page.module.scss";
import { generateMetaDataFromPage } from "@/lib/metadata/generate-metadata";
import { initialiseServices } from "@/lib/services/intialise-services";

type Params = Promise<{ route: string[] }>;
type Props = {
	params: Params;
};

export const revalidate = 960;

// export async function generateStaticParams() {
// 	return [
// 		{ route: ["world"] },
// 		{ route: ["us"] },
// 		{ route: ["uk"] },
// 		{ route: ["ukraine"] },
// 		{ route: ["trade-war"] },
// 	];
// }

export const generateMetadata = async ({ params }: Props) => {
	await initialiseServices();
	const { route } = await params;
	const joined = route.join("/");

	const metadata = await generateMetaDataFromPage(`/${joined}`);
	return metadata;
};

export default async function Page({ params }: Props) {
	await initialiseServices();
	const { route } = await params;
	const joined = route.join("/");

	const page = (await getPage(`/${joined}`)) as IPage;

	if (!page) {
		// omg do something
		return <h1>404</h1>;
	}

	return (
		<div className={styles.page}>
			<PageDisplay page={page} />
		</div>
	);
}
