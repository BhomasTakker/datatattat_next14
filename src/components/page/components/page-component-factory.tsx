import { PageContent } from "@/types/page";
import {
	PageContainersMap,
	PageComponentsOptions,
} from "./page-component-factory-map";

export const PageComponentFactory = ({ content }: { content: PageContent }) => {
	const { container } = content || {};
	const { containerType } = container;

	const ContainerComponent = PageContainersMap.get(
		containerType as PageComponentsOptions
	);

	// do better error handling
	if (!ContainerComponent) {
		return <div>Component not found</div>;
	}

	return <ContainerComponent content={content} />;
};
