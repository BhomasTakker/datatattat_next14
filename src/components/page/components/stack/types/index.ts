import { PageStackCollectionVariants } from "@/types/components/page/stack";

export type VerticalStackProps = {
	variant: PageStackCollectionVariants.Vertical;
};

export const ContainerWidth = {
	XXS: "180",
	XS: "240",
	SM: "320",
	MD: "640",
	LG: "768",
	XL: "1024",
	XXL: "1280",
} as const;

export const ContainerWidthOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerWidth = (typeof ContainerWidth)[keyof typeof ContainerWidth];

export type ContainerWidthOptions = keyof typeof ContainerWidth;

export const ContainerHeight = {
	XXS: "240",
	XS: "360",
	SM: "480",
	MD: "600",
	LG: "720",
	XL: "840",
	XXL: "960",
} as const;

export const ContainerHeightOptions = {
	XXS: "XXS",
	XS: "XS",
	SM: "SM",
	MD: "MD",
	LG: "LG",
	XL: "XL",
	XXL: "XXL",
} as const;

type ContainerHeight = (typeof ContainerHeight)[keyof typeof ContainerHeight];

type ContainerHeightOptions = keyof typeof ContainerHeight;
export const containerWidthOptions: ContainerWidthOptions[] = Object.keys(
	ContainerWidth
) as ContainerWidthOptions[];

// Row configuration for row-stack layout
export type Row = {
	columns: number;
	evenColumns: boolean;
	// you wouldn't max/min height - you would set a height
	maxHeight: ContainerHeightOptions;
	minWidth: ContainerWidthOptions;
	// these are all actually strings!
	// Unless we can/want to convert these after http requests use strings
	index?: string; // optional for default row - specifies which row this config applies to
};

export type Column = {
	minWidth: ContainerWidthOptions;
	maxHeight: ContainerHeightOptions;
	evenColumns: boolean;
};

export type RowStackProps = {
	variant: PageStackCollectionVariants.RowStack;
	defaultRow: Row; // Default configuration used for any row not specifically configured
	rows?: Row[]; // Optional array of row-specific configurations - index property determines which row each config applies to
};

export type PageStackProps = {
	variant: PageStackCollectionVariants;
} & (VerticalStackProps | RowStackProps);
