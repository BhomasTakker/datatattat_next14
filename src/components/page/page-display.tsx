import { IPage } from "@/types/page";
import { PageProfile } from "./profile/page-profile";
import { PageComponentFactory } from "./components/page-component-factory";
import styles from "./page-display.module.scss";
import { getCurrentRoute } from "@/utils/route";

export const PageDisplay = async ({ page }: { page: IPage }) => {
	const { content, meta, profile } = page || {};

	// not here
	const currentRoute = await getCurrentRoute();
	const isUserRoute = currentRoute?.includes("/users/") || false;
	const isClient = isUserRoute;

	return (
		<main className={styles.page}>
			{/* REPLACE with meta function */}
			{/* Page head for meta data, seo, etc */}
			{/* <PageHead headData={meta} /> */}
			{/* Page profile data for title, description, createdBy, etc */}
			<PageProfile profile={profile} />
			<PageComponentFactory content={content} isClient={isClient} />
		</main>
	);
};
