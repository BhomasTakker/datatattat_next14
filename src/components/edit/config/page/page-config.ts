import { GenericInput } from "@/types/edit/inputs/inputs";
import { EditInputs } from "../../inputs/inputs";
import { COMPONENT_CONFIG } from "../components/component-config";
import { PAGE_GRID_CONFIG } from "./components/page-grid";
import { PAGE_STACK_CONFIG } from "./components/stack/page-stack";

enum pageContainers {
	STACK = "Stack",
	GRID = "Grid",
}

type pageContainersProps = typeof PAGE_STACK_CONFIG | typeof PAGE_GRID_CONFIG;

const pageContainersMap = new Map<string, pageContainersProps>([
	[pageContainers.STACK, PAGE_STACK_CONFIG],
	[pageContainers.GRID, PAGE_GRID_CONFIG],
]);

export const PAGE_CONFIG: GenericInput = {
	// id used will be content
	id: "page",
	type: EditInputs.inputList,
	label: "Page",

	inputs: [
		{
			// add button here to save as page template - no we would want to save meta and page properties too no?
			// add button here to demo page
			id: "title",
			type: EditInputs.title,
			title: "Add Page interaction buttons here",
		},
		{
			id: "containerType",
			type: EditInputs.objectSelect,
			label: "Page Container",
			defaultValue: pageContainers.STACK,
			required: true,
			options: [pageContainers.STACK],
			optionMap: pageContainersMap,
			// save to/create on props object
			optionId: "props",
		},
		COMPONENT_CONFIG,
	],
};
