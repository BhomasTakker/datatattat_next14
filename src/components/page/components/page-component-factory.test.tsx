import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { PageComponentFactory } from "./page-component-factory";
import { PageContent } from "@/types/page";
import { PageComponentsOptions } from "./page-component-factory-options";

jest.mock("./stack/page-stack", () => {
	return {
		__esModule: true,
		PageStack: ({ content }: { content: PageContent }) => {
			const { container } = content || {};
			const { containerType } = container;
			return (
				<div data-testid="test-page-stack-component">{`Page Component ${containerType}`}</div>
			);
		},
	};
});

jest.mock("./grid/page-grid", () => {
	return {
		__esModule: true,
		PageGrid: ({ content }: { content: PageContent }) => {
			const { container } = content || {};
			const { containerType } = container;
			return (
				<div data-testid="test-page-grid-component">{`Page Component ${containerType}`}</div>
			);
		},
	};
});

const content = {
	container: {
		containerType: PageComponentsOptions.STACK,
		initData: {},
	},
	props: {},
	components: [],
};

describe("PageComponentFactory", () => {
	it("should render the Page Stack component", () => {
		render(<PageComponentFactory content={content} />);
		const component = screen.getByTestId("test-page-stack-component");
		expect(component).toBeInTheDocument();
	});

	it("should render the Page Grid component", () => {
		content.container.containerType = PageComponentsOptions.GRID;
		render(<PageComponentFactory content={content} />);
		const component = screen.getByTestId("test-page-grid-component");
		expect(component).toBeInTheDocument();
	});

	it("should render a default component if the container type is not found", () => {
		// @ts-expect-error - We are testing the default component
		content.container.containerType = "Unknown";

		render(<PageComponentFactory content={content} />);
		const component = screen.getByText("Component not found");
		expect(component).toBeInTheDocument();
	});

	it("renders PageComponentFactory unchanged", () => {
		content.container.containerType = PageComponentsOptions.STACK;
		const { container } = render(<PageComponentFactory content={content} />);
		expect(container).toMatchSnapshot();
	});
});
