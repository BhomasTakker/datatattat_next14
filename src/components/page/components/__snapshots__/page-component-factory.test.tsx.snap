// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageComponentFactory renders PageComponentFactory unchanged 1`] = `
<div>
  <div
    data-testid="test-page-stack-component"
  >
    Page Component Stack
  </div>
</div>
`;
